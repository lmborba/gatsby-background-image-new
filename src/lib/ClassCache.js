const componentClassCache = Object.create({});
/**
 * Cache component classes as we never know if a Component wasn't already set.
 *
 * @param className   string  className given by props
 * @return {*|boolean}
 */
export const inComponentClassCache = className => {
  return componentClassCache[className] || false;
};

/**
 * Adds a component's classes to componentClassCache.
 *
 * @param className   string  className given by props
 */
export const activateCacheForComponentClass = className => {
  if (className) {
    componentClassCache[className] = true;
  }
};

/**
 * Resets the componentClassCache (especially important for reliable tests).
 */
export const resetComponentClassCache = () => {
  for (const className in componentClassCache)
    delete componentClassCache[className];
};
