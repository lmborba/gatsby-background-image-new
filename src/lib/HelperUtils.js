import filterInvalidDOMProps from 'filter-invalid-dom-props';
import { groupByMedia, hasArtDirectionSupport } from './MediaUtils';

/**
 * Strip BackgroundImage propTypes from remaining props to be passed to <Tag />
 *
 * @param props
 * @return {Object}
 */
export const stripRemainingProps = props => filterInvalidDOMProps(props);

/**
 * Preprocess art directed images.
 *
 * @param props
 * @return {Object}
 */
export const convertProps = props => {
  const convertedProps = { ...props };
  const { fixed, fluid } = convertedProps;

  // if (fluid && !hasImageArray(props)) {
  //   convertedProps.fluid = [].concat(fluid)
  // }
  //
  // if (fixed && !hasImageArray(props)) {
  //   convertedProps.fixed = [].concat(fixed)
  // }

  // convert fluid & fixed to arrays so we only have to work with arrays
  if (fluid && hasArtDirectionSupport(props, 'fluid')) {
    convertedProps.fluid = groupByMedia(convertedProps.fluid);
  }
  if (fixed && hasArtDirectionSupport(props, 'fixed')) {
    convertedProps.fixed = groupByMedia(convertedProps.fixed);
  }

  return convertedProps;
};
