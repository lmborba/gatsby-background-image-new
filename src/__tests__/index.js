import { render } from '@testing-library/react';

import React from 'react';
import {
  fixedShapeMock,
  fluidShapeMock,
  setupBackgroundImage,
} from './mocks/Various.mock';
import BackgroundImage from '..';
import { activateCacheForImage, resetImageCache } from '../lib/ImageCache';
import { resetComponentClassCache } from '../lib/ClassCache';

afterEach(resetImageCache);

global.console.debug = jest.fn();

let elements = [];

// jest.mock('short-uuid')

describe(`<BackgroundImage />`, () => {
  const observe = jest.fn(callback => elements.push(callback));
  const unobserve = jest.fn(callback => elements.unshift(callback));
  const tmpRnd = Math.random;
  beforeEach(() => {
    global.IntersectionObserver = jest.fn(() => ({
      observe,
      unobserve,
    }));
    resetImageCache();
    // Mock Math.random beforehand, lest another random classname is created.
    Math.random = jest.fn(() => 0.424303425546642);
    // Freeze StyleUtils#fixClassName.
    const uuid = require('short-uuid');
    uuid.generate.mockImplementation(() => '73WakrfVbNJBaAmhQtEeDv');
  });
  afterEach(() => {
    Math.random = tmpRnd;
    resetComponentClassCache();
  });

  it(`should render fixed size images`, () => {
    const component = setupBackgroundImage({});
    expect(component).toMatchSnapshot();
  });

  it(`should render stacked fixed size images`, () => {
    const component = setupBackgroundImage({ multiImage: true });
    expect(component).toMatchSnapshot();
  });

  it(`should call critical fixed images`, () => {
    activateCacheForImage({ fixed: fixedShapeMock });
    const options = {
      addClass: true,
      critical: true,
      fixed: true,
    };
    const component = setupBackgroundImage(options);
    expect(component).toMatchSnapshot();
  });

  it(`should call critical images`, () => {
    resetImageCache();
    const options = {
      addClass: true,
      critical: true,
    };
    const component = setupBackgroundImage(options);
    expect(component).toMatchSnapshot();
  });

  it(`should call multiple critical fixed images`, () => {
    activateCacheForImage({ fixed: [fixedShapeMock, fixedShapeMock] });
    const options = {
      addClass: true,
      critical: true,
      fixed: true,
      multiImage: true,
    };
    const component = setupBackgroundImage(options);
    expect(component).toMatchSnapshot();
  });

  it(`should render fluid images`, () => {
    const component = setupBackgroundImage({ fluid: true });
    expect(component).toMatchSnapshot();
  });

  it(`should render without opacity hack with truthy preserveStackingContext`, () => {
    const component = setupBackgroundImage({
      fluid: true,
      props: { preserveStackingContext: true },
    });
    expect(component).toMatchSnapshot();
  });

  it(`should render multiple fluid images`, () => {
    const component = setupBackgroundImage({ fluid: true, multiImage: true });
    expect(component).toMatchSnapshot();
  });

  it(`should call critical fluid images`, () => {
    activateCacheForImage({ fluid: fluidShapeMock });
    const options = {
      fluid: true,
      addClass: true,
      critical: true,
    };
    const component = setupBackgroundImage(options);
    expect(component).toMatchSnapshot();
  });

  it(`should call multiple critical fluid images`, () => {
    activateCacheForImage({ fluid: [fluidShapeMock, fluidShapeMock] });
    const options = {
      fluid: true,
      addClass: true,
      critical: true,
      multiImage: true,
    };
    const component = setupBackgroundImage(options);
    expect(component).toMatchSnapshot();
  });

  it(`should work without external class`, () => {
    const { container } = render(
      <BackgroundImage fluid={fluidShapeMock}>
        <h1>testempty</h1>
      </BackgroundImage>
    );
    expect(container).toMatchSnapshot();
  });

  it(`should work without fluid / fixed image when keepStatic is given`, () => {
    const { container } = render(
      <BackgroundImage keepStatic>
        <h1 id="kept-child">I am a child</h1>
      </BackgroundImage>
    );
    expect(container).toMatchSnapshot();
  });

  it(`should change style.display from 'inherit' to 'inline-block'`, () => {
    const { container } = render(
      <BackgroundImage fixed={fixedShapeMock} style={{ display: `inherit` }}>
        <h1>testempty</h1>
      </BackgroundImage>
    );
    expect(container).toMatchSnapshot();
  });

  it(`should work with another external class`, () => {
    activateCacheForImage({ fluid: fluidShapeMock });
    const options = {
      fluid: true,
      addClass: true,
      additionalClass: `test`,
    };
    const component = setupBackgroundImage(options);
    expect(component).toMatchSnapshot();
  });

  it(`should return null without fluid / fixed`, () => {
    const options = {
      fixedClass: false,
      onLoad: null,
      onError: null,
      onStartLoad: null,
      fixed: false,
    };
    const component = setupBackgroundImage(options);
    expect(component).toMatchSnapshot();
  });

  it(`should work with className`, () => {
    const options = {
      fluid: true,
      additionalClass: `test`,
      fixed: false,
      addBackgroundColor: false,
    };
    const component = setupBackgroundImage(options);
    expect(component).toMatchSnapshot();
  });

  it(`should work with BackgroundColor`, () => {
    const options = {
      fluid: true,
      fixedClass: false,
      fixed: false,
      addBackgroundColor: `#fff`,
    };
    const component = setupBackgroundImage(options);
    expect(component).toMatchSnapshot();
  });

  it(`should work with fadeIn`, () => {
    const options = {
      fluid: true,
      fixedClass: false,
      fixed: false,
      fadeIn: true,
    };
    const component = setupBackgroundImage(options);
    expect(component).toMatchSnapshot();
  });

  it(`should create style tag with pseudo-elements`, () => {
    const options = {
      additionalClass: `test`,
    };
    const component = setupBackgroundImage(options);
    const styleTag = component.querySelector(`style`);
    expect(styleTag).toBeInTheDocument();
    expect(styleTag).toMatchSnapshot();
  });

  it(`should create style tag at fixed with pseudo-elements`, () => {
    const options = {
      addClass: true,
      additionalClass: `test`,
    };
    const component = setupBackgroundImage(options);
    const styleTag = component.querySelector(`style`);
    expect(styleTag).toBeInTheDocument();
    expect(styleTag).toMatchSnapshot();
  });

  it(`should have class with pseudo element in style tag`, () => {
    const options = {
      fluid: true,
      addClass: true,
      additionalClass: `test`,
    };
    const component = setupBackgroundImage(options);
    const styleTag = component.querySelector(`style`);
    expect(styleTag).toHaveTextContent(/background-repeat: repeat-y/g);
    expect(styleTag).toHaveTextContent(/background-position: center/g);
    expect(styleTag).toHaveTextContent(/background-size: contain/g);
  });
});
