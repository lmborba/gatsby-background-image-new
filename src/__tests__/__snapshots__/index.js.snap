// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BackgroundImage /> should call critical fixed images 1`] = `
<div>
  <div
    alt="Alt text for the image"
    class="imageClass  gbi-1393017994-73WakrfVbNJBaAmhQtEeDv"
    id="testid"
    style="background-position: center; background-repeat: repeat-y; background-size: contain; position: relative; display: inline-block; opacity: 0.99; width: 100px; height: 100px;"
    title="Title for the image"
  >
    <style>
      
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before,
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            background-color: lightgray;
            transition: none;
            background-position: center;
background-repeat: repeat-y;
background-size: contain;

          }
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before {
            z-index: -100;
            background-image: url('http://localhost/test_fixed_image.jpg');
            background-image: url(string_of_base64);
            opacity: 0; 
          }
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            z-index: -101;
            background-image: url('http://localhost/test_fixed_image.jpg');
            
            opacity: 1;
          }
        
    </style>
    <h1>
      test
    </h1>
  </div>
</div>
`;

exports[`<BackgroundImage /> should call critical fluid images 1`] = `
<div>
  <div
    alt="Alt text for the image"
    class="imageClass  gbi-1393017994-73WakrfVbNJBaAmhQtEeDv"
    id="testid"
    style="background-position: center; background-repeat: repeat-y; background-size: contain; position: relative; display: inline; opacity: 0.99;"
    title="Title for the image"
  >
    <style>
      
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before,
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            background-color: lightgray;
            transition: none;
            background-position: center;
background-repeat: repeat-y;
background-size: contain;

          }
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before {
            z-index: -100;
            background-image: url('http://localhost/test_fluid_image.jpg');
            background-image: url(string_of_base64);
            opacity: 0; 
          }
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            z-index: -101;
            background-image: url('http://localhost/test_fluid_image.jpg');
            
            opacity: 1;
          }
        
    </style>
    <h1>
      test
    </h1>
  </div>
</div>
`;

exports[`<BackgroundImage /> should call critical images 1`] = `
<div>
  <div
    alt="Alt text for the image"
    class="imageClass  gbi-1393017994-73WakrfVbNJBaAmhQtEeDv"
    id="testid"
    style="background-position: center; background-repeat: repeat-y; background-size: contain; position: relative; display: inline-block; opacity: 0.99; width: 100px; height: 100px;"
    title="Title for the image"
  >
    <style>
      
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before,
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            background-color: lightgray;
            transition: opacity 0.5s ease 250ms;
            background-position: center;
background-repeat: repeat-y;
background-size: contain;

          }
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before {
            z-index: -100;
            background-image: url(string_of_base64);
            
            opacity: 1; 
          }
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            z-index: -101;
            
            background-image: url(string_of_base64);
            
          }
        
    </style>
    <h1>
      test
    </h1>
  </div>
</div>
`;

exports[`<BackgroundImage /> should call multiple critical fixed images 1`] = `
<div>
  <div
    alt="Alt text for the image"
    class="imageClass  gbi-1393017994-73WakrfVbNJBaAmhQtEeDv"
    id="testid"
    style="background-position: center; background-repeat: repeat-y; background-size: contain; position: relative; display: inline-block; opacity: 0.99;"
    title="Title for the image"
  >
    <style>
      
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before,
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            background-color: lightgray;
            transition: none;
            background-position: center;
background-repeat: repeat-y;
background-size: contain;

          }
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before {
            z-index: -100;
            background-image: url(string_of_base64),url(string_of_base64);
            background-image: url(string_of_base64),url(string_of_base64);
            opacity: 0; 
          }
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            z-index: -101;
            background-image: url(string_of_base64),url(string_of_base64);
            
            opacity: 1;
          }
        
    </style>
    <h1>
      test
    </h1>
  </div>
</div>
`;

exports[`<BackgroundImage /> should call multiple critical fluid images 1`] = `
<div>
  <div
    alt="Alt text for the image"
    class="imageClass  gbi-1393017994-73WakrfVbNJBaAmhQtEeDv"
    id="testid"
    style="background-position: center; background-repeat: repeat-y; background-size: contain; position: relative; display: inline; opacity: 0.99;"
    title="Title for the image"
  >
    <style>
      
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before,
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            background-color: lightgray;
            transition: opacity 0.5s ease 250ms;
            background-position: center;
background-repeat: repeat-y;
background-size: contain;

          }
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before {
            z-index: -100;
            background-image: url(string_of_base64),url(string_of_base64),linear-gradient(to right, rgba(30, 75, 115, 1), rgba(255, 255, 255, 0));
            
            opacity: 1; 
          }
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            z-index: -101;
            
            background-image: url(string_of_base64),url(string_of_base64),linear-gradient(to right, rgba(30, 75, 115, 1), rgba(255, 255, 255, 0));
            
          }
        
    </style>
    <h1>
      test
    </h1>
  </div>
</div>
`;

exports[`<BackgroundImage /> should change style.display from 'inherit' to 'inline-block' 1`] = `
<div>
  <div
    class="gbi-1393017994-73WakrfVbNJBaAmhQtEeDv"
    style="background-position: center; background-repeat: no-repeat; background-size: cover; position: relative; opacity: 0.99; width: 100px; height: 100px;"
  >
    <style>
      
          .gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before,
          .gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            
            transition: opacity 0.5s ease 250ms;
            background-position: center;
background-repeat: no-repeat;
background-size: cover;

          }
          .gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before {
            z-index: -100;
            background-image: url(string_of_base64);
            
            opacity: 1; 
          }
          .gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            z-index: -101;
            
            background-image: url(string_of_base64);
            
          }
        
    </style>
    <h1>
      testempty
    </h1>
  </div>
</div>
`;

exports[`<BackgroundImage /> should create style tag at fixed with pseudo-elements 1`] = `
<style>
  
          .imageClass.test.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before,
          .imageClass.test.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            background-color: lightgray;
            transition: opacity 0.5s ease 250ms;
            background-position: center;
background-repeat: repeat-y;
background-size: contain;

          }
          .imageClass.test.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before {
            z-index: -100;
            background-image: url(string_of_base64);
            
            opacity: 1; 
          }
          .imageClass.test.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            z-index: -101;
            
            background-image: url(string_of_base64);
            
          }
        
</style>
`;

exports[`<BackgroundImage /> should create style tag with pseudo-elements 1`] = `
<style>
  
          .imageClass.test.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before,
          .imageClass.test.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            background-color: lightgray;
            transition: opacity 0.5s ease 250ms;
            background-position: center;
background-repeat: repeat-y;
background-size: contain;

          }
          .imageClass.test.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before {
            z-index: -100;
            background-image: url(string_of_base64);
            
            opacity: 1; 
          }
          .imageClass.test.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            z-index: -101;
            
            background-image: url(string_of_base64);
            
          }
        
</style>
`;

exports[`<BackgroundImage /> should render fixed size images 1`] = `
<div>
  <div
    alt="Alt text for the image"
    class="imageClass  gbi-1393017994-73WakrfVbNJBaAmhQtEeDv"
    id="testid"
    style="background-position: center; background-repeat: no-repeat; background-size: cover; position: relative; display: inline-block; opacity: 0.99; width: 100px; height: 100px;"
    title="Title for the image"
  >
    <style>
      
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before,
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            background-color: lightgray;
            transition: opacity 0.5s ease 250ms;
            background-position: center;
background-repeat: no-repeat;
background-size: cover;

          }
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before {
            z-index: -100;
            background-image: url(string_of_base64);
            
            opacity: 1; 
          }
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            z-index: -101;
            
            background-image: url(string_of_base64);
            
          }
        
    </style>
    <h1>
      test
    </h1>
  </div>
</div>
`;

exports[`<BackgroundImage /> should render fluid images 1`] = `
<div>
  <div
    alt="Alt text for the image"
    class="imageClass  gbi-1393017994-73WakrfVbNJBaAmhQtEeDv"
    id="testid"
    style="background-position: center; background-repeat: repeat-y; background-size: contain; position: relative; display: inline; opacity: 0.99;"
    title="Title for the image"
  >
    <style>
      
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before,
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            background-color: lightgray;
            transition: opacity 0.5s ease 250ms;
            background-position: center;
background-repeat: repeat-y;
background-size: contain;

          }
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before {
            z-index: -100;
            background-image: url(string_of_base64);
            
            opacity: 1; 
          }
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            z-index: -101;
            
            background-image: url(string_of_base64);
            
          }
        
    </style>
    <h1>
      test
    </h1>
  </div>
</div>
`;

exports[`<BackgroundImage /> should render multiple fluid images 1`] = `
<div>
  <div
    alt="Alt text for the image"
    class="imageClass  gbi-1393017994-73WakrfVbNJBaAmhQtEeDv"
    id="testid"
    style="background-position: center; background-repeat: repeat-y; background-size: contain; position: relative; display: inline; opacity: 0.99;"
    title="Title for the image"
  >
    <style>
      
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before,
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            background-color: lightgray;
            transition: opacity 0.5s ease 250ms;
            background-position: center;
background-repeat: repeat-y;
background-size: contain;

          }
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before {
            z-index: -100;
            background-image: url(string_of_base64),url(string_of_base64),linear-gradient(to right, rgba(30, 75, 115, 1), rgba(255, 255, 255, 0));
            
            opacity: 1; 
          }
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            z-index: -101;
            
            background-image: url(string_of_base64),url(string_of_base64),linear-gradient(to right, rgba(30, 75, 115, 1), rgba(255, 255, 255, 0));
            
          }
        
    </style>
    <h1>
      test
    </h1>
  </div>
</div>
`;

exports[`<BackgroundImage /> should render stacked fixed size images 1`] = `
<div>
  <div
    alt="Alt text for the image"
    class="imageClass  gbi-1393017994-73WakrfVbNJBaAmhQtEeDv"
    id="testid"
    style="background-position: center; background-repeat: no-repeat; background-size: cover; position: relative; display: inline-block; opacity: 0.99;"
    title="Title for the image"
  >
    <style>
      
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before,
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            background-color: lightgray;
            transition: opacity 0.5s ease 250ms;
            background-position: center;
background-repeat: no-repeat;
background-size: cover;

          }
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before {
            z-index: -100;
            background-image: url(string_of_base64),url(string_of_base64);
            
            opacity: 1; 
          }
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            z-index: -101;
            
            background-image: url(string_of_base64),url(string_of_base64);
            
          }
        
    </style>
    <h1>
      test
    </h1>
  </div>
</div>
`;

exports[`<BackgroundImage /> should render without opacity hack with truthy preserveStackingContext 1`] = `
<div>
  <div
    alt="Alt text for the image"
    class="imageClass  gbi-1393017994-73WakrfVbNJBaAmhQtEeDv"
    id="testid"
    style="background-position: center; background-repeat: repeat-y; background-size: contain; position: relative; display: inline; opacity: 0.99;"
    title="Title for the image"
  >
    <style>
      
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before,
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            background-color: lightgray;
            transition: opacity 0.5s ease 250ms;
            background-position: center;
background-repeat: repeat-y;
background-size: contain;

          }
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before {
            z-index: -100;
            background-image: url(string_of_base64);
            
            opacity: 1; 
          }
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            z-index: -101;
            
            background-image: url(string_of_base64);
            
          }
        
    </style>
    <h1>
      test
    </h1>
  </div>
</div>
`;

exports[`<BackgroundImage /> should return null without fluid / fixed 1`] = `<div />`;

exports[`<BackgroundImage /> should work with BackgroundColor 1`] = `
<div>
  <div
    alt="Alt text for the image"
    class="gbi-1393017994-73WakrfVbNJBaAmhQtEeDv"
    id="testid"
    style="background-position: center; background-repeat: no-repeat; background-size: cover; position: relative; display: inline; opacity: 0.99;"
    title="Title for the image"
  >
    <style>
      
          .gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before,
          .gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            background-color: #fff;
            transition: opacity 0.5s ease 250ms;
            background-position: center;
background-repeat: no-repeat;
background-size: cover;

          }
          .gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before {
            z-index: -100;
            background-image: url(string_of_base64);
            
            opacity: 1; 
          }
          .gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            z-index: -101;
            
            background-image: url(string_of_base64);
            
          }
        
    </style>
    <h1>
      test
    </h1>
  </div>
</div>
`;

exports[`<BackgroundImage /> should work with another external class 1`] = `
<div>
  <div
    alt="Alt text for the image"
    class="imageClass test gbi-1393017994-73WakrfVbNJBaAmhQtEeDv"
    id="testid"
    style="background-position: center; background-repeat: repeat-y; background-size: contain; position: relative; display: inline; opacity: 0.99;"
    title="Title for the image"
  >
    <style>
      
          .imageClass.test.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before,
          .imageClass.test.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            background-color: lightgray;
            transition: none;
            background-position: center;
background-repeat: repeat-y;
background-size: contain;

          }
          .imageClass.test.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before {
            z-index: -100;
            background-image: url('http://localhost/test_fluid_image.jpg');
            background-image: url(string_of_base64);
            opacity: 0; 
          }
          .imageClass.test.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            z-index: -101;
            background-image: url('http://localhost/test_fluid_image.jpg');
            
            opacity: 1;
          }
        
    </style>
    <h1>
      test
    </h1>
  </div>
</div>
`;

exports[`<BackgroundImage /> should work with className 1`] = `
<div>
  <div
    alt="Alt text for the image"
    class="imageClass test gbi-1393017994-73WakrfVbNJBaAmhQtEeDv"
    id="testid"
    style="background-position: center; background-repeat: repeat-y; background-size: contain; position: relative; display: inline; opacity: 0.99;"
    title="Title for the image"
  >
    <style>
      
          .imageClass.test.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before,
          .imageClass.test.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            
            transition: opacity 0.5s ease 250ms;
            background-position: center;
background-repeat: repeat-y;
background-size: contain;

          }
          .imageClass.test.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before {
            z-index: -100;
            background-image: url(string_of_base64);
            
            opacity: 1; 
          }
          .imageClass.test.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            z-index: -101;
            
            background-image: url(string_of_base64);
            
          }
        
    </style>
    <h1>
      test
    </h1>
  </div>
</div>
`;

exports[`<BackgroundImage /> should work with fadeIn 1`] = `
<div>
  <div
    alt="Alt text for the image"
    class="gbi-1393017994-73WakrfVbNJBaAmhQtEeDv"
    id="testid"
    style="background-position: center; background-repeat: no-repeat; background-size: cover; position: relative; display: inline; opacity: 0.99;"
    title="Title for the image"
  >
    <style>
      
          .gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before,
          .gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            background-color: lightgray;
            transition: opacity 0.5s ease 250ms;
            background-position: center;
background-repeat: no-repeat;
background-size: cover;

          }
          .gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before {
            z-index: -100;
            background-image: url(string_of_base64);
            
            opacity: 1; 
          }
          .gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            z-index: -101;
            
            background-image: url(string_of_base64);
            
          }
        
    </style>
    <h1>
      test
    </h1>
  </div>
</div>
`;

exports[`<BackgroundImage /> should work without external class 1`] = `
<div>
  <div
    class="gbi-1393017994-73WakrfVbNJBaAmhQtEeDv"
    style="background-position: center; background-repeat: no-repeat; background-size: cover; position: relative; opacity: 0.99;"
  >
    <style>
      
          .gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before,
          .gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            
            transition: opacity 0.5s ease 250ms;
            background-position: center;
background-repeat: no-repeat;
background-size: cover;

          }
          .gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before {
            z-index: -100;
            background-image: url(string_of_base64);
            
            opacity: 1; 
          }
          .gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            z-index: -101;
            
            background-image: url(string_of_base64);
            
          }
        
    </style>
    <h1>
      testempty
    </h1>
  </div>
</div>
`;

exports[`<BackgroundImage /> should work without fluid / fixed image when keepStatic is given 1`] = `
<div>
  <div
    class="gbi-2113735767-73WakrfVbNJBaAmhQtEeDv"
    style="background-position: center; background-repeat: no-repeat; background-size: cover; position: relative; opacity: 0.99;"
  >
    <style>
      
          .gbi-2113735767-73WakrfVbNJBaAmhQtEeDv:before,
          .gbi-2113735767-73WakrfVbNJBaAmhQtEeDv:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            
            transition: opacity 0.5s ease 250ms;
            background-position: center;
background-repeat: no-repeat;
background-size: cover;

          }
          .gbi-2113735767-73WakrfVbNJBaAmhQtEeDv:before {
            z-index: -100;
            
            
            opacity: 1; 
          }
          .gbi-2113735767-73WakrfVbNJBaAmhQtEeDv:after {
            z-index: -101;
            
            
            
          }
        
    </style>
    <h1
      id="kept-child"
    >
      I am a child
    </h1>
  </div>
</div>
`;
